src/signup/signup.service.ts

```diff
@@ -18,27 +18,43 @@ export class SignupService {
     private readonly jwtService: JwtService
   ) {}
 
-  private getConfirmationUrl(email: string, redirectLink: string) {
+  private createOriginFromHost(host: string) {
+    let scheme = "https";
+    if (host.startsWith("localhost")) {
+      scheme = "http";
+    }
+    return `${scheme}://${host}`;
+  }
+
+  private getConfirmationUrl(
+    email: string,
+    host: string,
+    redirectLink: string
+  ) {
     const userJwt = this.jwtService.sign(
       { email },
       {
         expiresIn: process.env.CONFIRMATION_JWT_EXPIRATION
       }
     );
-    const confirmationLink = `${process.env.GATEKEEPER_DOMAIN}${this.confirmationPath}?user=${userJwt}&r=${redirectLink}`;
+    const confirmationLink = `${this.createOriginFromHost(host) ||
+      process.env.DEFAULT_GATEKEEPER_ORIGIN}${
+      this.confirmationPath
+    }?user=${userJwt}&r=${redirectLink}`;
```

Comment:
```suggestion
    const confirmationLink = `${this.createOriginFromHost(host) ||
      process.env.DEFAULT_GATEKEEPER_ORIGIN}${
      this.confirmationPath
    }/?user=${userJwt}&r=${redirectLink}`;
```

As it currently stands, createOriginFromHost creates the url without the trailing slash (so it would be like `http://localhost:3000`, so the final URL with queryparams is `http://localhost:3000?user=foo&r=bar`). While this might be handled by some browsers it still is a little bit more risky to not add the trailing slash. Please make sure that the URLs that are generated are valid ones for all the uses of the ValidatedHost.

Reply:
That's a good catch. I'll make the needed changes

Reply:
@phulsechinmay Oh I see!